libgalois/src/analytics/bfs/bfs.cpp

```diff
@@ -246,9 +271,127 @@ SynchronousAlgo(
   }
 }
 
+template <bool CONCURRENT, typename P>
+void
+SynchronousDirectOpt(
+    Graph* graph, katana::PropertyGraph* transpose_graph_topo,
+    Graph::Node source, const P& pushWrap) {
+  using Cont = typename std::conditional<
+      CONCURRENT, katana::InsertBag<Graph::Node>,
+      katana::SerStack<Graph::Node>>::type;
+  using Loop = typename std::conditional<
+      CONCURRENT, katana::DoAll, katana::StdForEach>::type;
+
+  katana::GAccumulator<uint32_t> work_items;
+
+  Loop loop;
+
+  katana::DynamicBitset front_bitset, next_bitset;
+  front_bitset.resize(graph->size());
+  next_bitset.resize(graph->size());
+
+  front_bitset.reset();
+  next_bitset.reset();
+
+  Cont* frontier = new Cont();
+  Cont* next_frontier = new Cont();
+
+  Dist next_level{0};
+  graph->GetData<BfsNodeDistance>(source) = 0U;
+
+  if (CONCURRENT) {
+    pushWrap(*next_frontier, source, "parallel");
+  } else {
+    pushWrap(*next_frontier, source);
+  }
+
+  work_items += 1;
+
+  int64_t edges_to_check = graph->num_edges();
+  int64_t scout_count = graph->edges(source).size();
+  uint32_t num_nodes = graph->num_nodes();
+  uint64_t old_num_work_items{0};
+
+  katana::GAccumulator<uint64_t> writes_pull, writes_push;
+  writes_pull.reset();
+  writes_push.reset();
+
+  while (!next_frontier->empty()) {
+    std::swap(frontier, next_frontier);
+    next_frontier->clear();
+    // TODO(lhc) need to do parameter if (scout_count > edges_to_check / alpha)
+    if (scout_count > edges_to_check / 15) {
+      WlToBitset(*frontier, front_bitset);
+      do {
+        ++next_level;
+        old_num_work_items = work_items.reduce();
+        work_items.reset();
+
+        loop(
+            katana::iterate(*transpose_graph_topo),
+            [&](const typename Graph::Node& dst) {
+              auto& ddata = graph->GetData<BfsNodeDistance>(dst);
+              if (ddata == BfsImplementation::kDistanceInfinity) {
+                for (auto e : graph->edges(dst)) {
```

Comment:
Shouldn't this transpose_graph_topo?

Rename transpose_graph_topo to transpose_graph everywhere.

Reply:
No. `Graph` is TypedPropertyGraph, and IIUC, this is additional wrapper objects to handle properties. It does not have transposing or making symmetric. So what I did is to just change topology. `PropertyGraph` has transposing function. I just called this, transposed the topology, still node data are on `TypedPropertyGraph`, and so, access only those data. Therefore, I could not change any of them.

Reply:
I understood that but it doesn't answer my question. You should be accessing in_edges here, which is edges of transposed_graph.

Don't use "topo". PropertyGraph has properties too.

Reply:
The additional wrapper in TypedPropertyGraph is to give a typed view of those properties.